File: java/src/main/java/com/xunfos/comparison/java/iv_extras/I_seq_sample.java
package com.xunfos.comparison.java.iv_extras;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class I_seq_sample {
    public static void main(String[] args) {
        var list = new ArrayList<Integer>();
        for (int i = 0; i <= 10; i++) {
            list.add(i);
        }

        final List<Integer> doubledAndPair = list.stream()
                .map(it -> {
                    System.out.println("mapping " + it);
                    return it * it;
                })
                .filter(it -> {
                    System.out.println("filtering " + it);
                    return it % 2 == 0;
                })
                .collect(Collectors.toList());

        doubledAndPair.forEach(System.out::println);
    }


}

File: kt/src/main/kotlin/com/xunfos/comparison/kt/iv_extras/i_seq_sample.kt
package com.xunfos.comparison.kt.iv_extras

import kotlinx.coroutines.flow.*
import kotlinx.coroutines.runBlocking

fun main() {
    sequenceCollection()
    coroutineCollection()
}

fun sequenceCollection() {
    val doubledAndPair = (1..10)
        .asSequence()
        .map {
            println("Mapping $it")
            it * it
        }
        .filter {
            println("Filtering $it")
            it % 2 == 0
        }

    doubledAndPair.forEach(::println)
}

fun coroutineCollection() {
    val doubledAndPair = (1..10)
        .asFlow()
        .map {
            println("Mapping $it")
            it * it
        }
        .filter {
            println("Filtering $it")
            it % 2 == 0
        }

    runBlocking {
        doubledAndPair.collect(::println)
    }
}
